### PR TITLE
Query exceptions inside transactions will not propage out when caught with try-catch block.

### LINKED ISSUE #95: Cannot throw custom exceptions in response to PostgreSQLException within transaction
Exceptions thrown in response to a `PostgreSQLException` in a transaction block are not propagated out. Instead, `transaction` throws the original `PostgreSQLException`.

A test that demonstrates this issue:

```dart
    test('exception thrown in transaction is propagated out', () async {
      final expectedException = Exception('my custom exception');
      dynamic actualException;
      await conn.transaction((c) async {
        await c.query('INSERT INTO t (id) VALUES (1)');
        try {
          await c.query('INSERT INTO t (id) VALUES (1)');
        } on PostgreSQLException catch (_) {
          throw expectedException;
        }
      }).catchError((error) => actualException = error);

      // Fails, this is actually the original PostgreSQLException
      expect(actualException, expectedException);
    });
```

This code is on a branch at https://github.com/bryanoltman/postgresql-dart/blob/custom-error/test/transaction_test.dart#L652-L665.

### PR DESCRIPTION
**Pull request by isoos**
Fixes #95.

### CODE REVIEW

File: lib/src/connection.dart [hunk 1]
```diff
@@ -639,7 +636,7 @@ abstract class _PostgreSQLExecutionContextMixin
     }
   }
 
-  Future _onQueryError(Query query, dynamic error, [StackTrace? trace]) async {}
+  // Future _onQueryError(Query query, dynamic error, [StackTrace? trace]) async {}
```
Comment (bryanoltman) [inline lib/src/connection.dart:639]: Remove commented-out code

File: test/transaction_test.dart [hunk 2]
```diff
@@ -648,5 +648,40 @@ void main() {
...
+    test('exception caught in transaction is not propagated out', () async {
+      await conn.transaction((c) async {
+        await c.query('INSERT INTO t (id) VALUES (1)');
+        try {
+          await c.query('INSERT INTO t (id) VALUES (1)');
+        } catch (_) {
+          // ignore
+        }
+      });
+    });
```
Comment (bryanoltman) [inline test/transaction_test.dart:685]: This seems incomplete – is it missing an `expect(…, completes);`?

File: test/transaction_test.dart [hunk 3]
```diff
@@ -648,5 +648,40 @@ void main() {
...
+      }).catchError((error) => actualException = error);
+      expect(actualException, expectedException);
+
+      // testing the same exception without the try-catch block inside the transaction:
+      dynamic uncaughtException;
+      await conn.transaction((c) async {
+        await c.query('INSERT INTO t (id) VALUES (1)');
+        await c.query('INSERT INTO t (id) VALUES (1)');
+      }).catchError((error) => uncaughtException = error);
+      expect(uncaughtException.toString(), thrownException.toString());
```
Comment (bryanoltman) [inline test/transaction_test.dart:673]: It might be worth making this a separate test with the comment as the title

**Comment by bryanoltman** [comment]
Thanks for the quick fix! Left a few comments.

File: test/transaction_test.dart (hunk 3 above)
Comment (isoos) [reply test/transaction_test.dart:673]: This tests the message with the thrown exception message from the previous block, it makes more sense to tie these together.

**Review by bryanoltman — APPROVED** [review]

**Comment by isoos** [comment]
@bryanoltman: It seems that we have a few tests that explicitly rely on the old behavior of having unawaited queries and throwing in the transaction. I'm having a second thought on the scope of the breaking behavior, and it seems that this may be better moved to the v3 version of the API. wdyt? also cc @simolus3 for opinions

**Comment by bryanoltman** [comment]
I think including this in a major release seems like a fine idea.

Also, hi Simon! 😄

**Comment by simolus3** [comment]
Hi Bryan :wave: 

> also cc @simolus3 for opinions

At the moment, we don't have any builtin transaction methods in the v3 API at all - users have to run the `BEGIN` and `COMMIT` statements manually for now. I agree that a transaction API based on callbacks should rethrow exceptions (or at least wrap them consistently and in a way that allows obtaining the original cause). I'm happy to start a draft for such API in the v3 library.

I don't really have a strong opinion on whether this should be changed in the v2 API since I'm not using it. But maybe we should find a path to eventually stabilize the v3 release instead ;)

**Comment by isoos** [comment]
Closing in favor of #174.

**Pull request closed without merging**